### DISCUSSION TITLE
KMP version of MapCompose

### DISCUSSION THREAD

**Discussion by p-lr** [peterLaurence/MapCompose#112] [Ideas]
Plans for the future KMP port of MapCompose

**Comment by p-lr** [comment]
@rkreienbuehl We can continue here if you don't mind.
If I understand correctly, you started from Feb 8 commit, which is quite recent. You may have taken the pool from MapComposeDesktop because it wasn't using Java's specific ConcurrentHashMap.
I will think about a way to refactor this without using a JVM specific type. If I'm successful, I'll push on master.

**Reply by rkreienbuehl** [reply]
Thats right, I changed the pool to use the one from MapComposeDesktop. Maybe a solution is to use the [Stately library](https://github.com/touchlab/Stately) from touchlab, but maybe theres also another solution. I take a look at this as soon as I am finished with porting the demo apps.

**Reply by p-lr** [reply]
I've done it on master branch.

**Reply by p-lr** [reply]
One important thing to note is that the pool on `master` requires we provide the bitmap size. The reason is the pool is actually a pool of pool. For example, if we want to get a Bitmap of size 1234, the returned Bitmap from the pool is taken from an internal "sub-pool" which stores Bitmaps of size 1234.
While on the other hand, the former pool used in MapComposeDesktop requires no argument.

**Reply by rkreienbuehl** [reply]
I will take a look at it and try to integrate it next week

**Comment by rkreienbuehl** [comment]
@p-lr I just pushed the port of the demo app on my repo. The app works for ios, desktop and android, but I had to change `ui/widgets/Callout.kt` a little bit and the design could be improved especially on desktop.

**Reply by p-lr** [reply]
Thanks! I just tested a couple of demos on my android device. For instance there's noticeable slowdowns in tile rendering (compared to the current android native version). Have you also seen it?

**Reply by p-lr** [reply]
I'm investigating the performance issue. Here is what I have for KMP, with HWUI turrned on:
<img src="https://github.com/p-lr/MapCompose/assets/15638794/12f53c70-9229-494b-b284-6f2036e25b03" width="200">
And this is on android native:
<img src="https://github.com/p-lr/MapCompose/assets/15638794/2acc5e54-ee60-49d4-9aea-acebfdae3a38" width="200">

**Reply by p-lr** [reply]
I believe I found the root cause. It will require some structural changes. For instance, I need to take the time to think about the strategy. 
Since I noticed a lot of differences in core classes such as TileCanvasState and TileCollector, this is what I suggest:
1. I create a `kmp` branch in MapCompose repo
2. I re-create the project structure you made in MapComposeKMP, and I just copy original files when relevant *without adding or editing anything* in the files
3. I explain how to fix the rendering issue, and then you start applying "kmp" changes to the code base (such as adding `expects` and `actuals`). This way, your contributions will be clearly visible.

What do you think?

**Reply by p-lr** [reply]
Alternatively to 3., I can do the kmp changes myself but I don't commit. I send you a patch file so that you commit the changes.

**Reply by rkreienbuehl** [reply]
This aproach is good for me. I can make a fork and do a pull request after changing the files.

I could only try the android version in the simulator. There it was about the same speed as the original android app, but the demos can maybe use some optimization and for the desktop app I want to add buttons for interaction.

**Reply by p-lr** [reply]
Yes, desktop demo need a way to zoom. Performance wise, I also thought initially about missing optimizations.
I'll tell you when I'm done.

**Reply by p-lr** [reply]
I have created a new repo called [MapComposeMP](https://github.com/p-lr/MapComposeMP).
While was applying your changes, I also made some adjustments and fixes. For example, paths are now working. As you can see in git history, I regrouped most of your changes in two commits (for the core lib and demo app, respectively).
I haven't tested the ios demo app, because I don't have a mac, nor and apple device. However now, I will probably buy those^^

For the ios demo app, the `Config.xconfig` will have to be updated. As I haven't developed a ios app yet, I'm not sure what to do.

**Reply by rkreienbuehl** [reply]
Thats great, I will take a look this at it evening. I also started porting the BitmapPool and maybe found a way to use `Bitmap` instead of `ByteArray`, maybe this way the code could be more like in mapcompose without so much difference, but I am not finished yet, so I cannot tell for sure.

**Reply by p-lr** [reply]
I removed the BitmapPool, on purpose. It's useless because we can't reuse existing memory and every time we get a tile we allocate a whole ByteArray. So don't bother ;)

**Reply by rkreienbuehl** [reply]
I made a fork today and tested the demos. All of them worked really well. I also made some changes to the demos. 
When I see it right, there are no ZoomIn/ZoomOut interactions by now in MapState?

> For the ios demo app, the `Config.xconfig` will have to be updated. As I haven't developed a ios app yet, I'm not sure what to do.

In `Config.xconfig` one has to change the`TEAM_ID`, which is the ID of the developer-team when registring at apple for a developer account. I also changed the `BUNDLE_ID` and `APP_NAME` in my last commit on the fork.

**Comment by p-lr** [comment]
> When I see it right, there are no ZoomIn/ZoomOut interactions by now in MapState?

I'm not sure I understand. What do you mean?

> I made a fork today and tested the demos. All of them worked really well. I also made some changes to the demos.

Great! Don't hesitate to make a PR.

**Reply by rkreienbuehl** [reply]
I made a PR.
I thought about interaction like `MapState.scrollTo()` for zoomIn and zoomOut, but I realized that I can use `scrollTo()` for that.

**Reply by p-lr** [reply]
Yes indeed, when using scrollTo, we can specify a relative position and also a scale.
